### DISCUSSION TITLE
Why is the marker size limit for vips2jpeg.c 65530?

### DISCUSSION THREAD

**Discussion by usualuse** [libvips/libvips#2490] [General]
I am developing an API server that resizes and saves images uploaded by our customers. In the process of operating and maintaining that API server, I came across a question.

In vips2jpeg.c, if the size of the marker exceeds 64KB, the marker will not be saved.
I would like to know why this limit is 65530.

I think the limit for not exceeding 64KB is 65534, except for the marker identifier and marker type.

In fact, there is an image uploaded by a customer whose exif data (FFE1) has a size of 0xFFFE.
Editing this image file with exiftool and Mac preview.app is no problem, but when I save it with libvips, the exif is deleted.

**Comment by jcupitt** [comment]
Hi @usualuse,

That's a good question, I'll have a look.

Do you have an image with an 0xfffe marker I could use for testing?

**Reply by usualuse** [reply]
Thanks for your reply.
Sorry, I only have one image with EXIF length 0xFFFE. Since it is your image, we cannot provide it as a sample.

**Reply by usualuse** [reply]
Hi @jcupitt ,

I can't provide customer images, but I was able to create an image with a similar EXIF length 0xFFFE.

When I save this image as a jpeg, I get the following message.

`VIPS: field "exif-data" is too large for a single JPEG marker, ignoring`

![510482](https://user-images.githubusercontent.com/24588932/138281930-d4ce1969-1971-4220-b9fd-0bd4ab807b67.jpg)

**Reply by jcupitt** [reply]
That's great, thank you! 

We should probably move further discussion to the enhancement issue.

**Comment by jcupitt** [comment]
The limit is here, with a comment:

https://github.com/libvips/libvips/blob/master/libvips/foreign/vips2jpeg.c#L264

libvips will split large ICC profiles up, but doesn't have any code to do this for other large metadata objects, like EXIF or XMP.

There's a comment on the difficulty of writing large XMP:

https://github.com/libvips/libvips/blob/master/libvips/foreign/vips2jpeg.c#L306

So I think the answer to your question is that reading and writing large EXIF data chunks would be a useful enhancement to libvips. Let's move this to an issue and tag it.

**Comment by jcupitt** [comment]
Enhancement issue: https://github.com/libvips/libvips/issues/2493
